### DISCUSSION TITLE
Bypass modifier (shift) on long press

### DISCUSSION THREAD

**Discussion by alloc33** [manna-harbour/miryoku#117] [Q&A] [answered]
I'm trying to find a way how to bypass a modifier key on long press. For example I want to hold "k" so it should be recognized as "kkkkkkkkkk...." How can I do that?

**Comment by manna-harbour** [comment, marked as answer]
In general the Base layer keys don't auto repeat, as many are tap hold. The thumbs have basic versions on the layers that can be used for auto repeat, such as backspace and delete on Nav (although for those, combining with control is generally preferable). For the alphas, the cursor keys on Nav can be used instead of holding for navigation in vi, and the Tap layer can be used for games etc.

However, ZMK supports tap hold auto repeat with https://zmk.dev/docs/behaviors/hold-tap#quick-tap-ms. Support for easily customising mods will be added in future, see https://github.com/manna-harbour/miryoku/issues/33. For now you'd need to manually add it here https://github.com/manna-harbour/miryoku_zmk/blob/095fd13c5c003008d33d001a81702ae1f9d7ba74/miryoku/miryoku.dtsi#L28-L35.

**Reply by alloc33** [reply]
> In general the Base layer keys don't auto repeat, as many are tap hold. The thumbs have basic versions on the layers that can be used for auto repeat, such as backspace and delete on Nav (although for those, combining with control is generally preferable). For the alphas, the cursor keys on Nav can be used instead of holding for navigation in vi, and the Tap layer can be used for games etc.
> 
> However, ZMK supports tap hold auto repeat with https://zmk.dev/docs/behaviors/hold-tap#quick-tap-ms. Support for easily customising mods will be added in future, see #33. For now you'd need to manually add it here https://github.com/manna-harbour/miryoku_zmk/blob/095fd13c5c003008d33d001a81702ae1f9d7ba74/miryoku/miryoku.dtsi#L28-L35.

Thanks for your answer, but I still don't understand how to switch between auto-repeat and tap-hold. Should it be done with additional features layer? If yes, how can I access this layer?

**Reply by manna-harbour** [reply]
You need to edit https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku.dtsi and add
```
      quick-tap-ms = <125>;
```
immediately after this line https://github.com/manna-harbour/miryoku_zmk/blob/095fd13c5c003008d33d001a81702ae1f9d7ba74/miryoku/miryoku.dtsi#L34. Then double tap within 125ms and hold the second tap to auto repeat.

But I recommend the previously mentioned alternatives instead.

**Reply by alloc33** [reply]
> 

Sorry, what do you mean under previously mentioned? Do you mean I can access to basic character layers through this? If yes - how? I tried but this doesn't seems to work (just switched to colemak somehow accidentally and had to reset my board:/)

![miryoku-kle-features](https://user-images.githubusercontent.com/45129790/183408344-a65057f4-073a-4ac4-8d51-4824ddc6959c.png)

**Reply by manna-harbour** [reply]
> Sorry, what do you mean under previously mentioned? 

I was just referring to the first paragraph here: https://github.com/manna-harbour/miryoku/discussions/117#discussioncomment-3345486. What is the use case? If it's for vi, you can instead use the cursor keys on Nav and the vi-style alternative layout option. See https://github.com/manna-harbour/miryoku/tree/master/docs/reference#vi-style. Although it's generally recommended to use the higher level movement commands in vi rather than holding hjkl. 

> Do you mean I can access to basic character layers through this? If yes - how? I tried but this doesn't seems to work (just switched to colemak somehow accidentally and had to reset my board:/)

You can use the Tap layer for games etc. By default, you need a hardware reset button to exit Tap. You'll probably want to customise the layer or the layer mapping anyway, in that case you can add a key to return to base (`&to U_BASE`). See https://github.com/manna-harbour/miryoku/discussions/85. To set Tap to QWERTY see https://github.com/manna-harbour/miryoku/tree/master/docs/reference#tap, i.e. use `#define MIRYOKU_TAP_QWERTY` in the custom_config workflow option https://github.com/manna-harbour/miryoku_zmk#custom_config or in the config file https://github.com/manna-harbour/miryoku_zmk#config-file.

**Reply by alloc33** [reply]
The use case is for Homerow app. For scrolling. Just figured out how to change their shortcuts) But thanks for help, now I know hop to use alt layers
